services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/HistoryPolicySpecValidator.java

```diff
@@ -0,0 +1,88 @@
+package com.linkedin.openhouse.tables.api.validator.impl;
+
+import com.linkedin.openhouse.common.api.spec.TableUri;
+import com.linkedin.openhouse.tables.api.spec.v0.request.components.History;
+import com.linkedin.openhouse.tables.api.spec.v0.request.components.TimeGranularity;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component
+public class HistoryPolicySpecValidator {
+
+  private String failureMessage = "";
+  private String errorField = "";
+
+  protected boolean validate(History history, TableUri tableUri) {
+    if (history != null) {
+      if (history.getMaxAge() <= 0 && history.getVersions() <= 0) {
+        failureMessage =
+            String.format(
+                "Must define either a time based retention or count based retention for snapshots in table %s",
+                tableUri);
+        return false;
+      }
+
+      if (history.getGranularity() == null && history.getMaxAge() > 0
+          || history.getGranularity() != null && history.getMaxAge() <= 0) {
+        failureMessage =
+            String.format(
+                "Incorrect maxAge specified. history.maxAge must be defined together with history.granularity for table %s",
+                tableUri);
+        return false;
+      }
+
+      if (!validateHistoryConfigMaxAgeWithinBounds(history)) {
+        failureMessage =
+            String.format(
+                "History for the table [%s] max age must be between 1 to 3 days", tableUri);
+        return false;
+      }
+
+      if (!validateHistoryConfigVersionsWithinBounds(history)) {
+        failureMessage =
+            String.format("History for the table [%s] must be between 2 to 100 versions", tableUri);
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Validate that the amount of time to retain history of table snapshots is between 1 to 3 days
+   * versions
+   *
+   * @param history
+   * @return
+   */
+  protected boolean validateHistoryConfigMaxAgeWithinBounds(History history) {
+    int maxAge = history.getMaxAge();
+    TimeGranularity granularity = history.getGranularity();
+    boolean isGranularityValid =
+        granularity == null
+            || granularity.equals(TimeGranularity.HOUR)
+            || granularity.equals(TimeGranularity.DAY);
+
+    boolean isMaxAgeValid =
+        granularity != null
+            && (maxAge < 3 && granularity.equals(TimeGranularity.DAY)
+                || maxAge < 72 && granularity.equals(TimeGranularity.HOUR))
+            && (maxAge > 1 && granularity.equals(TimeGranularity.DAY)
+                || maxAge > 24 && granularity.equals(TimeGranularity.HOUR));
+
+    return isGranularityValid && isMaxAgeValid;
+  }
+
+  protected boolean validateHistoryConfigVersionsWithinBounds(History history) {
```

Comment:
Should we add some comments on why we choose this? Or are we planning to override this method?

Reply:
I'll leave a comment to explain why at least 2 (always have a backup available) but we can override this too.